File: verification/src/main/java/org/msr/mnr/verification/expressions/IteExpr.java
package org.msr.mnr.verification.expressions;

import org.msr.mnr.verification.utils.Packet;

public class IteExpr extends Expr {
    private static final long serialVersionUID = 7L;

    private BoolExpr condition;

    public IteExpr(BoolExpr condition, Object left, Object right) {
        super(left,right);
        this.condition = condition;
    }

    @Override
    public int[] evaluate(Packet p) {
        // System.out.println("input packet: ");
        // System.out.println(p);
        boolean branch = condition.evaluate(p);
        // System.out.print("branch output: ");
        // System.out.println(branch);
        if (branch) {
            // System.out.println("going in left");
            return evalSingle(left, leftType, p);
        } else {
            // System.out.println("going in right");
            return evalSingle(right, rightType, p);
        }
    }


    @Override
    public String toString() {
        return "( ITE " + condition.toString() + " ? " + left.toString() + " : " + right.toString() + ")";
    }
}

File: C++Verifier/src/expressions/IteExpr.cpp
#ifndef ITEEXPR_H
#define ITEEXPR_H

#include <vector>
// #include "../utils/Packet.cpp"
#include "BoolExpr.cpp"

using namespace std;

class IteExpr: public Expr {
public:
    IteExpr(shared_ptr<BoolExpr> _condition, shared_ptr<Expr> left, shared_ptr<Expr> right) : Expr(move(left), move(right)) {
        condition = move(_condition);
    }

    void* evaluate(shared_ptr<Packet> p) {

        bool branch = (!!static_cast<int>(reinterpret_cast<std::uintptr_t>(condition -> evaluate(p))));
        string* ret = nullptr;
        if (branch) {
            *ret = evalSingle(left, leftType, p);
        } else {
            *ret = evalSingle(right, rightType, p);
        }
        return static_cast<void*>(ret);
    }

    string toString() {
        return "( ITE " + (condition -> toString()) + " ? " + (left -> toString()) + " : " + (right -> toString()) + ")";
    }

private:
    shared_ptr<BoolExpr> condition;
};

#endif